subsys/bluetooth/services/bas_c.c

```diff
@@ -140,7 +253,10 @@ int bt_gatt_bas_c_handles_assign(struct bt_gatt_dm *dm,
 	}
 	LOG_DBG("Getting handles from battery service.");
 
-	bt_gatt_bas_c_init(bas_c);
+	/* If connection is established again, cancel previous read request. */
+	k_delayed_work_cancel(&bas_c->periodic_read.read_work);
```

Comment:
`k_delayed_work_cancel` called on an item that is not submitted is undefined.
```
	if (atomic_test(&bas_c->periodic_read.process, BAS_PERIODIC_READ_PROCESS)) {
		k_delayed_work_cancel(&bas_c->periodic_read.read_work);
	}
```

Reply:
`k_delayed_work_cancel ` called on an item that is not submitted do nothing. I know you mean that sentence in workqueue documentation.

> 
> The result of calling this on a k_delayed_work item that has
> not been submitted (i.e. before the return of the
> k_delayed_work_submit_to_queue() call) is undefined.

But all workqueue function used common spinlock instance so one function cannot be interrupted from other contex before return. I analyzed source code and history of this doc comment. This sentence was added because when you call cancel before submit. Submit will do despite this cancel call .